src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -248,17 +259,72 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
     }
   }
 
-  /*
   function performHighPriWork() {
-    // There is no such thing as high pri work yet.
+    // Always start from the root
+    nextUnitOfWork = findNextUnitOfWork();
+    while (nextUnitOfWork &&
+           nextPriorityLevel !== NoWork) {
+      nextUnitOfWork = performUnitOfWork(nextUnitOfWork);
+      if (!nextUnitOfWork) {
+        // Keep searching for high pri work until there's no more left
+        nextUnitOfWork = findNextUnitOfWork();
+      }
+      // Stop if the next unit of work is low priority
+      if (nextPriorityLevel > HighPriority) {
+        scheduleLowPriCallback(performLowPriWork);
+        return;
+      }
+    }
   }
 
-  function ensureHighPriIsScheduled() {
-    scheduleHighPriCallback(performHighPriWork);
+  function scheduleHighPriWork(root: FiberRoot, priorityLevel : PriorityLevel) {
+    // Set the priority on the root, without deprioritizing
+    if (root.current.pendingWorkPriority === NoWork ||
+        priorityLevel <= root.current.pendingWorkPriority) {
+      root.current.pendingWorkPriority = priorityLevel;
+    }
+
+    if (root.isScheduled) {
+      // If we're already scheduled, we can bail out.
+      return;
+    }
+    root.isScheduled = true;
+    if (lastScheduledRoot) {
+      // Schedule ourselves to the end.
+      lastScheduledRoot.nextScheduledRoot = root;
+      lastScheduledRoot = root;
+    } else {
+      // We're the only work scheduled.
+      nextScheduledRoot = root;
+      lastScheduledRoot = root;
+      scheduleHighPriCallback(performHighPriWork);
+    }
+  }
+
+  function scheduleWork(root : FiberRoot) {
+    if (defaultPriority === NoWork) {
+      return;
+    }
+    if (defaultPriority >= LowPriority) {
```

Comment:
This should actually be `defaultPriority >= HighPriority`.

We ended up overloading the term which needs to be undone. We should rename `scheduleHighPriWork` to `scheduleAnimationWork` because `HighPriority` is the highest priority that goes in to `requestIdleCallback`. We should only do `AnimationPriority` work during `requestAnimationFrame`.

Additionally, `SynchronousPriority` should not use `scheduleHighPriWork` so it should throw in that case.